go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -0,0 +1,510 @@
+		// disable redo logging and double write buffer if we are configured to do so.
+		err = params.Mysqld.ExecuteSuperQuery(ctx, "ALTER INSTANCE DISABLE INNODB REDO_LOG")
+		if err != nil {
+			return nil, vterrors.Wrap(err, "unable to disable REDO_LOG")
+		}
+		params.Logger.Infof("Disabled REDO_LOG")
+
+		defer func() { // re-enable once we are done with the restore.
+			err := params.Mysqld.ExecuteSuperQuery(ctx, "ALTER INSTANCE ENABLE INNODB REDO_LOG")
+			if err != nil {
+				params.Logger.Errorf("unable to re-enable REDO_LOG: %v", err)
+			} else {
+				params.Logger.Infof("Disabled REDO_LOG")
+			}
+		}()
+	}
+
+	// we need to disable SuperReadOnly otherwise we won't be able to restore the backup properly.
+	// once the backups is complete, we will restore it to its previous state.
+	resetFunc, err := be.handleSuperReadOnly(ctx, params)
+	if err != nil {
+		return nil, vterrors.Wrap(err, "unable to disable super-read-only")
+	}
+	defer resetFunc()
+
+	args := []string{}
+
+	if mysqlShellFlags != "" {
+		args = append(args, strings.Fields(mysqlShellFlags)...)
+	}
+
+	args = append(args, "-e", fmt.Sprintf("util.loadDump(%q, %s)",
+		bm.BackupLocation,
+		mysqlShellLoadFlags,
+	))
+
+	cmd := exec.CommandContext(ctx, "mysqlsh", args...)
+
+	params.Logger.Infof("running %s", cmd.String())
+
+	cmdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, vterrors.Wrap(err, "cannot create stdout pipe")
+	}
+	cmdErr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, vterrors.Wrap(err, "cannot create stderr pipe")
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, vterrors.Wrap(err, "can't start xbstream")
+	}
+
+	cmdWg := &sync.WaitGroup{}
+	cmdWg.Add(2)
+	go scanLinesToLogger(mysqlShellBackupEngineName+" stdout", cmdOut, params.Logger, cmdWg.Done)
+	go scanLinesToLogger(mysqlShellBackupEngineName+" stderr", cmdErr, params.Logger, cmdWg.Done)
+	cmdWg.Wait()
+
+	// Get the exit status.
+	if err := cmd.Wait(); err != nil {
+		return nil, vterrors.Wrap(err, mysqlShellBackupEngineName+" failed")
+	}
+	params.Logger.Infof("%s completed successfully", mysqlShellBackupBinaryName)
+
+	// disable local_infile now that the restore is done.
+	err = params.Mysqld.ExecuteSuperQuery(ctx, "SET GLOBAL LOCAL_INFILE=0")
+	if err != nil {
+		return nil, vterrors.Wrap(err, "unable to set local_infile=0")
+	}
+	params.Logger.Infof("set local_infile=0")
+
+	params.Logger.Infof("Restore completed")
+
+	return &bm.BackupManifest, nil
+}
+
+// ShouldDrainForBackup satisfies the BackupEngine interface
+// MySQL Shell backups can be taken while MySQL is running so we can control this via a flag.
+func (be *MySQLShellBackupEngine) ShouldDrainForBackup(req *tabletmanagerdatapb.BackupRequest) bool {
+	return mysqlShellBackupShouldDrain
+}
+
+func (be *MySQLShellBackupEngine) backupPreCheck(location string) error {
+	if mysqlShellBackupLocation == "" {
+		return fmt.Errorf("%w: no backup location set via --mysql_shell_location", MySQLShellPreCheckError)
```

Comment:
This error needs to use the new flag format (`--mysql-shell-backup-location`)

```suggestion
		return fmt.Errorf("%w: no backup location set via --mysql-shell-backup-location", MySQLShellPreCheckError)
```